### ISSUE TITLE
How to use CertificateProviders?

### ISSUE THREAD

**Issue by hognevevle** [ebekker/ACMESharp#216]
Hi,

I'm writing a C# application using ACMESharp, but I'm hitting a wall when trying to use the `OpenSslLibProvider`. 

This is what I'm using (as seen in [letsencrypt-win-simple](https://github.com/Lone-Coder/letsencrypt-win-simple/blob/master/letsencrypt-win-simple/Program.cs#L747)):
`var cp = CertificateProvider.GetProvider();`

At this point, I'm getting a `TargetInvocationException`.

` System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidOperationException: unresolved architecture-specific implementation  
at ACMESharp.PKI.Providers.OpenSslLibProvider..ctor(IReadOnlyDictionary'2 newParams) in C:\\projects\\acmesharp\\ACMESharp\\ACMESharp\\PKI\\Providers\\OpenSslLibProvider.cs:line 450 `

I've inspected `OpenSslLibProvider.cs`, and see that it checks for a 64 bits process. System.Environment.Is64BitProcess evaluates to `true` in my application.

My project currently references:
- ACMESharp
- ACMESharp.PKI.Providers.OpenSslLib64
- ManagedOpenSsl64

(I previously referenced both 64 bits and 32 bits assemblies, but I've now removed the 32 bits ones)

Does anybody have any clues on how I should resolve this? :)

**Comment by hognevevle** [comment]
I gave it a shot to use the BouncyCastleProvider instead, and now getting another error. I've removed all the OpenSsl assemblies, and added ACMESharp.PKI.Providers.BouncyCastle.dll (built from source), as well as BouncyCastle.Crypto (not sure if that one is needed).

`System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at ACMESharp.PKI.CertificateProvider.GetProvider(String name, IReadOnlyDictionary'2 initParams) i C:\\Users\\Hogne\\Downloads\\ACMESharp-master\\ACMESharp\\ACMESharp\\PKI\\CertificateProvider.cs:line 188`

**Comment by hognevevle** [comment]
Managed to make it work by using
```
CertificateProvider.RegisterProvider<BouncyCastleProvider>("BouncyCastle"); 
var cp = CertificateProvider.GetProvider("BouncyCastle");
```

Perhaps that is the way it is intended to work? If so, it would be nice to have it documented somewhere.

**Comment by gdau** [comment]
I know this is a bit old, but I had the same issue and solved it by manually adding a reference to the ACMESharp.PKI.Providers.OpenSslLib32 dll from my consuming application (in your case, the web app); for some reason it's not picked up automatically from the class library and it's not added via nuget (for me anyway).

So make sure that you add the reference from the packages folder where your class library is referencing the ACMESharp.PKI.Providers.OpenSslLib32 from.

To recap, the application that consumes my class lib has references to:

ACMESharp.PKI.Providers.OpenSslLib64
ACMESharp.PKI.Providers.OpenSslLib32
ManagedOpenSsl
ManagedOpenSsl64

Started working for me after that :)

**Comment by ohadschn** [comment]
The workaround suggested by @gdau worked for me too, but it resulted in another error:
`Unabled to create Azure Web Site Management client System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.TypeInitializationException: The type initializer for 'ACMESharp.PKI.Providers.OpenSslLibBaseProvider' threw an exception. ---> System.TypeInitializationException: The type initializer for 'OpenSSL.Core.Native' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'x86\libeay32': The specified module could not be found. (Exception from HRESULT: 0x8007007E)`

The workaround for that was pretty simple though, I just set `libeay32.dll` (and `ssleay32.dll` for good measure) to be copied to the output directory (_Copy if newer_).

**Comment by ebekker** [comment]
The OpenSSL library approach is deprecated and will be removed in the future because of the fragile dependency requirements with the native libraries.  Going forward, BouncyCastle is the recommended approach.

**Comment by ebekker** [comment]
@hognevevle, the *clean* way of resolving a CertificateProvider is using code like this:

```csharp
using (var pki = PkiToolExtManager.GetPkiTool("ProviderName"))
{
    // use pki which is IPkiTool implementation which calls
    // into provider-specific implementation
    var rsaKeyParams = new RsaPrivateKeyParams();
    var rsaKey = pki.GeneratePrivateKey(rsaKeyParams);
}
```

There is not a lot of documentation for using the ACMESharp client library as yet, so other consumers of this lib have just been using the samples in the unit tests to figure it out, sorry.

**Comment by ebekker** [comment]
On caveat, the code sample above is for the new method introduced in 0.8.2.
